### PR TITLE
Group by key

### PR DESCRIPTION
**Pull request by coszio**
## Progress
- [x] internals
- [x] unit tests
- [x] integration tests
- [x] define specific API
- [x] REST
- [x] gRPC

### All Submissions:

* [x] Have you followed the guidelines in our Contributing document?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../../pulls) for the same update/change?

### New Feature Submissions:

1. [x] Does your submission pass tests?
2. [x] Have you lint your code locally using `cargo fmt` command prior to submission?
3. [x] Have you checked your code using `cargo clippy` command?

### CODE REVIEW

**Comment by agourlay** [comment]
@coszio I see you are making some progress here :+1: 

One concern though is that it looks you are going to go over 1k lines of changes in this PR which is getting big to review.

Maybe you break the PR in two:
1. finding the groups
2. filling up the incomplete groups which is rather complex

WDYT?

**Comment by coszio** [comment]
@agourlay it's a good idea, I can split the aggregator part into a separate PR, so that it is easier to look at

File: lib/segment/src/types.rs [hunk 1]
```diff
@@ -197,6 +198,13 @@ impl PartialEq for ScoredPoint {
     }
 }
 
+impl Hash for ScoredPoint {
```
Comment (generall) [inline lib/segment/src/types.rs:201]: rm

**Review by ffuugoo — CHANGES_REQUESTED** [review]
Did a _partial_ review. Some minor remarks, but PR looks good so far. Will finish the review tomorrow.

**Comment by ffuugoo** [comment]
@coszio I've kinda already implemented all the trivial changes, I'll push them on a separate branch in a few minutes, so you can take a look and merge them into your branch.

File: lib/collection/src/grouping/aggregator.rs [hunk 2]
```diff
@@ -0,0 +1,519 @@
...
+
+        Ok(())
+    }
+
+    /// Adds multiple points to the group that they corresponds based on the group_by field, assumes that the points always have the grouped_by field, else it just ignores them
+    pub(super) fn add_points(&mut self, points: &[ScoredPoint]) {
+        points
+            .iter()
+            .map(|point| self.add_point(point))
+            .any(|r| matches!(r, Err(AllGroupsFull)));
```
[the remark this answers is not in the document]
Comment (coszio) [reply lib/collection/src/grouping/aggregator.rs:75]: Yeah, the intention is to do an early stop if we already have enough results, so the `any` here is just acting as circuit breaker

[the remark this answers is not in the document]
Comment (ffuugoo) [reply lib/collection/src/grouping/aggregator.rs:75]: Oooof. I get it now. A simple loop with explicit `break` would be much more readable. 🙈

**Comment by coszio** [comment]
@ffuugoo Really good points on the review so far :raised_hands:.I'm actually in the middle of a refactor after a live review from @generall feat. @agourlay so it's good that yours was only *partial*.

**Comment by ffuugoo** [comment]
> I'll push them on a separate branch in a few minutes...

@coszio https://github.com/qdrant/qdrant/tree/grouping-by-review

File: lib/collection/src/grouping/aggregator.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (timvisee) [reply lib/collection/src/grouping/aggregator.rs:65]: You can also use [`take_while`](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.take_while) here. But I'm not sure how much better that is.

[the remark this answers is not in the document]
Comment (coszio) [reply lib/collection/src/grouping/aggregator.rs]: Went ahead with the simple for loop with an `if let`, since the `take_while` suggests that we should pipe the output somewhere too

File: lib/collection/src/grouping/types.rs [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+            Value::Number(n) => n.hash(state),
+            Value::String(s) => s.hash(state),
+            _ => unreachable!("GroupKey should only be a number or a string"),
+        }
+    }
+}
+
+/// Abstraction over ScoredPoint to be used in a HashSet
+#[derive(Eq, Debug, PartialEq, Clone)]
+pub(super) struct HashablePoint(pub ScoredPoint);
```
[the remark this answers is not in the document]
Comment (coszio) [reply lib/collection/src/grouping/types.rs]: I actually need to be able to hash points with vectors and payloads in the final stage of the `group_by()` to avoid doing extra loops in the hydration part. Btw, I refactored the `hydrate_from` into the `Group` implementation.

tldr: I did the inner ScoredPoint private

File: lib/api/build.rs [hunk 4]
```diff
@@ -143,11 +143,17 @@ fn configure_validation(builder: Builder) -> Builder {
             ("SearchPoints.vector_name", "custom = \"crate::grpc::validate::validate_not_empty\""),
             ("SearchBatchPoints.collection_name", "length(min = 1, max = 255)"),
             ("SearchBatchPoints.search_points", ""),
+            // ("SearchPointGroups.search", "required"), // <-- TODO: needs Serialize on the underlying type
```
Comment (agourlay) [inline lib/api/build.rs:146]: What about those two TODOs?

File: lib/collection/src/grouping/types.rs [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+impl Hash for HashablePoint {
+    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
+        self.0.id.hash(state);
+        self.0.version.hash(state);
+    }
+}
+
+impl Ord for HashablePoint {
+    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
+        self.0.score.partial_cmp(&other.0.score).unwrap()
```
Comment (agourlay) [inline lib/collection/src/grouping/types.rs:115]: why the `unwrap` here?

File: openapi/tests/openapi_integration/test_group.py [hunk 6]
```diff
@@ -0,0 +1,135 @@
+import pytest
+
+from .helpers.helpers import request_with_validation
+from .helpers.collection_setup import basic_collection_setup, drop_collection
+
+collection_name = 'test_collection_reco'
```
Comment (agourlay) [inline openapi/tests/openapi_integration/test_group.py:6]: maybe it would be good to have name which mention grouping

File: openapi/tests/openapi_integration/test_group.py [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+
+def upsert_chunked_docs(collection_name, docs=50, chunks=5):
+    points = []
+    for doc in range(docs):
+        for chunk in range(chunks):
+            doc_id = f"doc_{doc}"
+            i = doc * chunks + chunk
+            p = {"id": i, "vector": [0.5, 0.5, 0.5, 0.5], "payload": {"docId": doc_id}}
```
Comment (agourlay) [inline openapi/tests/openapi_integration/test_group.py:14]: what do you think about introducing a few points where docID as different types:
- numbers
- string
- arrays

As a sanity check that the API is handling heterogeneous grouping key.

File: openapi/tests/openapi_integration/test_group.py [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+        api='/collections/{collection_name}/points/search/groups',
+        method="POST",
+        path_params={'collection_name': collection_name},
+        body={
+            "search": {
+                "vector": [1.0, 0.0, 0.0, 0.0],
+                "limit": 10,
+                "with_payload": True,
+            },
+            "group_by": "docId",
```
Comment (agourlay) [inline openapi/tests/openapi_integration/test_group.py:47]: would be nice to have a case where the group_by key is not existent to validate that we get empty groups

File: openapi/tests/openapi_integration/test_group.py [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+            "top": 3,
```
Comment (agourlay) [inline openapi/tests/openapi_integration/test_group.py:48]: I have doubt about the naming here, `top` vs `limit`.

We have been aliasing `top` to `limit` in the search related APIs.
Because it  works well with a potential `offset` parameters.

I wonder if we should just use `limit` in the group API.

It also clarify the relation between inner limit and outer limit :thinking:

File: lib/collection/src/grouping/aggregator.rs [hunk 10]
```diff
@@ -0,0 +1,425 @@
...
+    }
+
+    // Returns the best groups sorted by their best hit. The hits are sorted too
+    pub(super) fn distill(&self) -> Vec<Group> {
+        self.groups
+            .iter()
+            .map(|(k, v)| {
+                let hits = v
+                    .iter()
+                    .sorted()
```
Comment (agourlay) [inline lib/collection/src/grouping/aggregator.rs:103]: I think the `sorted` + reversing pattern could be optimized.
 
`sorted` requires to allocates the all collection

Might as well use `sorted_by` to sort it directly in descending order to skip the `reverse` iterator.

File: lib/collection/src/grouping/aggregator.rs [hunk 11]
```diff
@@ -0,0 +1,425 @@
...
+                    .collect::<Vec<_>>();
+
+                Group {
+                    hits,
+                    key: k.clone(),
+                    group_by: self.grouped_by.clone(),
+                }
+            })
+            // TODO: what happens when the best groups are not the most filled ones?
+            .sorted_by_key(|group| group.hits.len())
```
Comment (agourlay) [inline lib/collection/src/grouping/aggregator.rs:116]: same comment about sort + rev

File: lib/collection/src/grouping/types.rs (hunk 5 above)
Comment (coszio) [reply lib/collection/src/grouping/types.rs:115]: My bad, I indended to use the ScoredPoint Ord instead of reimplementing it

File: openapi/tests/openapi_integration/test_group.py (hunk 7 above)
Comment (coszio) [reply openapi/tests/openapi_integration/test_group.py:14]: Great idea, will do

File: openapi/tests/openapi_integration/test_group.py (hunk 6 above)
Comment (coszio) [reply openapi/tests/openapi_integration/test_group.py:6]: Copy-pasted :see_no_evil:

File: lib/collection/src/grouping/aggregator.rs [hunk 12]
```diff
@@ -0,0 +1,425 @@
...
+                    .cloned()
+                    .collect::<Vec<_>>();
+
+                Group {
+                    hits,
+                    key: k.clone(),
+                    group_by: self.grouped_by.clone(),
+                }
+            })
+            // TODO: what happens when the best groups are not the most filled ones?
```
Comment (agourlay) [inline lib/collection/src/grouping/aggregator.rs:115]: I believe we should select groups to returned according to their max scores and not according to their size.

File: lib/api/build.rs (hunk 4 above)
Comment (coszio) [reply lib/api/build.rs:146]: I wanted to ask if validating it manually on the tonic interface would be ok, since it doesn't seem I can just add `required` here. Has anyone faced this problem before?

File: lib/collection/src/grouping/aggregator.rs [hunk 13]
```diff
@@ -0,0 +1,425 @@
...
+            self.full_groups.insert(group_key);
+        }
+
+        Ok(())
+    }
+
+    /// Adds multiple points to the group that they corresponds based on the group_by field, assumes that the points always have the grouped_by field, else it just ignores them
+    pub(super) fn add_points(&mut self, points: &[ScoredPoint]) {
+        for point in points {
+            if let Err(AllGroupsFull) = self.add_point(point) {
```
Comment (agourlay) [inline lib/collection/src/grouping/aggregator.rs:62]: When matching against enum, I personally prefer exhaustive match for clarity and safety.

Something like

```rust
match self.add_point(point) {
   case Err(AllGroupsFull) => break,
   case Err(KeyNotFound | A | B) => continue,
   case Ok(_) => continue,
}
```

File: lib/collection/src/grouping/aggregator.rs (hunk 10 above)
Comment (coszio) [reply lib/collection/src/grouping/aggregator.rs:132]: Took me a little to get it. Both `sorted` and `sorted_by` consume the whole iterator, and both say
> The sorted iterator, if directly collected to a Vec, is converted without any extra copying or allocation cost.

I'm not sure if there is any difference here, but we can skip the `rev` usage if it helps, `rev` only traverses the iterator in reverse anyway.

File: lib/collection/src/grouping/aggregator.rs (hunk 12 above)
Comment (coszio) [reply lib/collection/src/grouping/aggregator.rs]: Fixed this using a priority queue and modifying the `group_by` method accordingly

File: lib/segment/src/types.rs [hunk 14]
```diff
@@ -197,6 +198,14 @@ impl PartialEq for ScoredPoint {
     }
 }
 
+#[derive(Debug, Serialize, Deserialize, JsonSchema, Clone)]
+pub struct PointGroup {
+    /// Scored points that have the same group_id
+    pub hits: Vec<ScoredPoint>,
+    /// Key shared by all the hits
+    pub group_id: serde_json::Map<String, Value>,
```
Comment (ffuugoo) [inline lib/segment/src/types.rs:206]: `serde_json::Map<String, Value>` is just to produce `{"<key>": "<value>"}` JSON, right? As a minor nitpick, I'd define a `struct GroupId { key: String, value: Value }` with a custom `Serialize` impl that would do that.

File: lib/collection/src/grouping/types.rs [hunk 15]
```diff
@@ -0,0 +1,184 @@
...
+}
+
+impl Hash for GroupKey {
+    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
+        match &self {
+            Self::Number(n) => n.hash(state),
+            Self::String(s) => s.hash(state),
+        }
+    }
+}
```
Comment (ffuugoo) [inline lib/collection/src/grouping/types.rs:83]: `#[derive(Hash)]`?

```suggestion
/// Abstraction over serde_json::Value to be used as a key in a HashMap/HashSet
#[derive(Debug, Eq, PartialEq, Clone, Hash)]
pub(super) enum GroupKey {
    String(String),
    Number(serde_json::Number),
}

impl TryFrom<serde_json::Value> for GroupKey {
    type Error = AggregatorError;

    /// Only allows Strings and Numbers to be converted into GroupKey
    /// When dealing with arrays, it will consider only the first element
    fn try_from(value: serde_json::Value) -> Result<Self, Self::Error> {
        let value = match value {
            serde_json::Value::Array(arr) => arr.into_iter().next().ok_or(KeyNotFound)?,
            _ => value,
        };
        match value {
            serde_json::Value::String(s) => Ok(Self::String(s)),
            serde_json::Value::Number(n) => Ok(Self::Number(n)),
            _ => Err(BadKeyType),
        }
    }
}

impl From<GroupKey> for serde_json::Value {
    fn from(key: GroupKey) -> Self {
        match key {
            GroupKey::String(s) => serde_json::Value::String(s),
            GroupKey::Number(n) => serde_json::Value::Number(n),
        }
    }
}
```

**Review by ffuugoo — COMMENTED** [review]
🚲

File: lib/collection/src/grouping/types.rs [hunk 16]
```diff
@@ -0,0 +1,133 @@
...
+    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
+        self.0.cmp(&other.0)
+    }
+}
+
+impl PartialOrd for HashablePoint {
+    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
+        self.0.partial_cmp(&other.0)
+    }
+}
```
[the remark this answers is not in the document]
Comment (coszio) [reply lib/collection/src/grouping/types.rs:127]: Interesting, does deriving Ord apply the inner Ord on the wrapper?

[the remark this answers is not in the document]
Comment (ffuugoo) [reply lib/collection/src/grouping/types.rs:127]: <details>
<summary>Yep.</summary>
Снимок экрана 2023-05-12 в 18 18 23
</details>

File: lib/segment/src/types.rs (hunk 14 above)
Comment (coszio) [reply lib/segment/src/types.rs:206]: The reason to use a serde_json::Map was to generate an `object` type on the JsonSchema, as the Value was generating something else. Ideally I'd prefer to skip repeating the group_by field in the result. Do you know a way to achieve this?

**Comment by generall** [comment]
todo: do not use `Number(serde_json::Number),` in public api

File: lib/segment/src/types.rs (hunk 14 above)
Comment (coszio) [reply lib/segment/src/types.rs]: nvm, already switched to a new GroupId struct

**Review by generall — APPROVED** [review]

File: lib/collection/Cargo.toml [hunk 17]
```diff
@@ -54,6 +54,7 @@ num_cpus = "1.15.0"
 tar = "0.4.38"
 fs_extra = "1.3.0"
 semver = "1.0.17"
+priority-queue = "1.3.1"
```
Comment (agourlay) [inline lib/collection/Cargo.toml:57]: I believe this dependency is now unused

File: lib/collection/src/grouping/aggregator.rs [hunk 18]
```diff
@@ -0,0 +1,466 @@
...
+        self.groups.len()
+    }
+
+    /// Return `max_groups` number of keys of the groups with the best score
+    fn best_group_keys(&self) -> Vec<&GroupId> {
+        match self.order {
+            Order::LargeBetter => {
+                self.group_max_scores
+                    .iter()
+                    .sorted_by_key(|(_, score)| OrderedFloat(**score))
```
Comment (agourlay) [inline lib/collection/src/grouping/aggregator.rs:125]: we could sort by the opposite comparator right away to avoid having to reverse the iterator

File: lib/segment/src/data_types/groups.rs [hunk 19]
```diff
@@ -0,0 +1,85 @@
...
+            GroupId::NumberI64(n) => json!(n),
+        }
+    }
+}
+
+impl TryFrom<&serde_json::Value> for GroupId {
+    type Error = ();
+
+    /// Only allows Strings and Numbers to be converted into GroupKey
+    /// When dealing with arrays, it will consider only the first element
```
Comment (agourlay) [inline lib/segment/src/data_types/groups.rs:51]: outdated comment about array first element

**Review by ffuugoo — APPROVED** [review]
LGTM! Terrific job, @coszio! 🔥
